railgun/internal/ctrlutils/utils.go

```diff
@@ -59,3 +67,79 @@ func IsAPIAvailable(mgr ctrl.Manager, obj client.Object) (bool, error) {
 }
+
+// HasAnnotation is a helper function to determine whether an object has a given annotation, and whether it's
+// to the value provided.
+func HasAnnotation(obj client.Object, key, expectedValue string) bool {
+	foundValue, ok := obj.GetAnnotations()[key]
+	return ok && foundValue == expectedValue
+}
+
+// MatchesIngressClassName indicates whether or not an object indicates that it's supported by the ingress class name provided.
+func MatchesIngressClassName(obj client.Object, ingressClassName string) bool {
+	if ing, ok := obj.(*netv1.Ingress); ok {
+		if ing.Spec.IngressClassName != nil && *ing.Spec.IngressClassName == ingressClassName {
+			return true
+		}
+	}
+	return HasAnnotation(obj, annotations.IngressClassKey, ingressClassName)
+}
+
+type objWithIngressClassNameSpec struct {
+	Spec struct{ IngressClassName *string }
+}
+
+// GeneratePredicateFuncsForIngressClassFilter builds a controller-runtime reconcilation predicate function which filters out objects
+// which do not have the "kubernetes.io/ingress.class" annotation configured and set to the provided value or in their .spec.
+func GeneratePredicateFuncsForIngressClassFilter(name string, specCheckEnabled, annotationCheckEnabled bool) predicate.Funcs {
+	preds := predicate.NewPredicateFuncs(func(obj client.Object) bool {
+		if annotationCheckEnabled && IsIngressClassAnnotationConfigured(obj, name) {
+			return true
+		}
+		if specCheckEnabled && IsIngressClassSpecConfigured(obj, name) {
+			return true
+		}
+		return false
+	})
+	preds.UpdateFunc = func(e event.UpdateEvent) bool {
+		if annotationCheckEnabled && IsIngressClassAnnotationConfigured(e.ObjectOld, name) || IsIngressClassAnnotationConfigured(e.ObjectOld, name) {
+			return true
+		}
+		if annotationCheckEnabled && IsIngressClassAnnotationConfigured(e.ObjectNew, name) || IsIngressClassAnnotationConfigured(e.ObjectNew, name) {
+			return true
+		}
+		return false
+	}
+	return preds
+}
+
+// IsIngressClassAnnotationConfigured determines whether an object has an ingress.class annotation configured that
+// matches the provide IngressClassName (and is therefore an object configured to be reconciled by that class).
+//
+// NOTE: keep in mind that the ingress.class annotation is deprecated and will be removed in a future release
+//       of Kubernetes in favor of the .spec based implementation.
+func IsIngressClassAnnotationConfigured(obj client.Object, expectedIngressClassName string) bool {
+	if foundIngressClassName, ok := obj.GetAnnotations()[annotations.IngressClassKey]; ok {
+		if foundIngressClassName == expectedIngressClassName {
+			return true
+		}
+	}
+	return false
+}
+
+// IsIngressClassAnnotationConfigured determines whether an object has IngressClassName field in its spec and whether the value
+// matches the provide IngressClassName (and is therefore an object configured to be reconciled by that class).
+func IsIngressClassSpecConfigured(obj client.Object, expectedIngressClassName string) bool {
```

Comment:
1. Consider using a simple type switch instead of reflection, for two reasons.
    - brevity: less code
    - performance: this is much faster than reflection
    - defensive programming: the current code has a nasty side effect that, when in 2 years from now, us or some other engineer (with no memory of this reflection trick) adds a new type to KIC, it will automatically interpret this field using its name by convention.
```go
switch obj := obj.(type) {
case netv1.Ingress:
  return obj.Spec.IngressClassName != nil && *obj.Spec.IngressClassName == expectedIngressClassName
}

return false
```

Reply:
Yep, this is a good call. `reflect` is fun, but costly. Changed in rebase :+1: